### DISCUSSION TITLE
Docker setup: Using the DOMAIN variable for verify email link URL

### DISCUSSION THREAD

**Discussion by nodje** [dani-garcia/vaultwarden#2462] [Q&A]
I'm using a docker setup behind a reverse proxy.
I've setup the DOMAIN variable and the HELO_NAME but the verify email link still comes with the http://localhost domain.

I'm not sure what else i should verify. login in the container and verifying the variable values shows everything is correct.

I'm running the container under a specific uid:gid, could this cause some issue in general?

**Comment by Masgalor** [comment]
I think it would help us to help you, if you could post the `support string` for your instance.

` admin Page -> Diagnostics `

**Comment by nodje** [comment]
ok, it took me too many trial and errors to understand what is happening.
ADMIN_TOKEN didn't seem to work anymore after I had set it to use docker-compose secrets.
And I had some weird behaviors, like changing the environments variables in docker-compose didn't seem to be picked up the way it should.

Now I get it: I mount the /data folder to a docker host backed-up folder, and once the `/data/config.json` has been created the values are set and whatever you put in docker-compose env variable is not taken into account as `/data/config.json` takes precedence.

Now I realize I haven't followed the doc to the letter.
But somewhere this should be explained.

In my case, if I want to be able to control the env variables that are stored in config.json from the docker-compose.yml file, and as such the config.json regenerated after each container recreation, I need to have the config.json stored outside of /data.

Or I control everything from the config.json file directly, and store this in my git repo. But then I can't use the secrets mecanism.

**Reply by nodje** [reply]
OK, re-reading the Overview part of the doc explains part of the problem.
But I got confused because I never changed anything though the ADMIN page, I just enabled it to SEE more variables.

And it's quite confusing to me that the ENV variables are not taking precedence over file config. 

So in my case, I need to delete the config.json, delete the ADMIN_TOKEN env variable and refer to https://github.com/dani-garcia/vaultwarden/blob/main/.env.template for other variable that could be configured.

**Comment by Masgalor** [comment]
> But somewhere this should be explained.

While I can understand that this might be not intuitive, it is the first paragraph in the wiki [section configuration](https://github.com/dani-garcia/vaultwarden/wiki/Configuration-overview) where this is explained.

The only values written to `config.json` are settings that were changed in the admin-panel.
If you dont touch any settings there, you can still control everything with env variables.
But as soon as you change something in the admin-panal it overrides the correspondig env variable.

**Reply by nodje** [reply]
Actually, AFAIU, that's not how it works exactly: once you enabled admin panel by setting the ADMIN_TOKEN env variable, everything gets written to config.json and then changes to existing ENV VAR are not taken into account anymore.

**Comment by BlackDex** [comment]
All settings are saved to the `config.json` file. But only when you click `save`. Not when enabling the `ADMIN_TOKEN`.

If that file exists it will overwrite all `ENV` values (as is mentioned at the top of the admin settings page).
Everything which is double underscored are settings set via `ENV` but also loaded via the `config.json` value.

So basically, if you have configured anything manually via the admin interface, then all the `ENV`'s are out the window, expect for a few read-only values which you can not change at runtime, like the PORT to listen on. LOG_LEVEL etc..

If  you have configured the `DOMAIN` as an environment variable, and later saved your config via the admin interface, and want to change the `DOMAIN` value, then you need to do that via the admin interface. Now i would suggest to also do that in your docker config, just to be sure in case you delete the config.json for example.

**Reply by nodje** [reply]
Correct, that's what happened exactly.
I remember now having tested saving a value through the admin page, which then cancelled the possibility of changing existing `ENV` var values.

So if I understand correctly I can actually activate admin page with ADMIN_TOKEN and keep configuring with `ENV` as long as I don't save anything through the admin page.
And if I need to test something handily and save some values with the Admin page I can always delete the config.json afterwards to restore the `ENV` config behavior.

**Comment by nodje** [comment]
Sadly though, this makes me realize that using docker-compose secrets in this fashion is not supported:
```
secrets:
  vaultwarden_smtp_password:
    file: ./vaultwarden_smtp_password.txt
  vaultwarden_admin_token:
    file: ./vaultwarden_admin_token.txt

services:
vaultwarden:
[...]
  environment:
      - ADMIN_TOKEN=/run/secrets/vaultwarden_admin_token #do not save values with it
      - SMTP_PASSWORD=/run/secrets/vaultwarden_smtp_password
  secrets:
      - vaultwarden_smtp_password
      - vaultwarden_admin_token
```

which defeats the purpose of using `ENV` in order to avoid comiting secrets to Git.

In this case, using config.json becomes more handy. I'm still reluctant to commit any password info, especially an ADMIN_TOKEN
